src/Categories/Multi/Category/Cartesian/Representable.agda

```diff
@@ -0,0 +1,174 @@
+open import Data.List.Membership.Propositional.Properties
+open import Data.List.Relation.Unary.Any
+import Data.List.Relation.Unary.Any.Properties as AnyP
+open import Data.Product
+open import Data.Sum
+open import Data.Wrap
+open import Function.Base as F using (case_return_of_)
+open import Level
+open import Relation.Binary.PropositionalEquality as ≡ using (_≡_)
+
+open import Categories.Category
+open import Categories.Category.Cartesian
+import Categories.Category.Cartesian.Properties as ×P
+open import Categories.Multi.Category.Cartesian
+
+private
+  variable
+    o ℓ e : Level
+
+module Extras {a} {A : Set a} where
+  -- TODO: contribute to stdlib
+
+  ∈-++⁺ : ∀ {x : A} {xs ys} (i : x ∈ xs ⊎ x ∈ ys) → x ∈ xs ++ ys
+  ∈-++⁺ = [ ∈-++⁺ˡ , ∈-++⁺ʳ _ ]′
+
+  ∈-++⁻∘++⁺ˡ : ∀ {x : A} {xs} ys (i : x ∈ xs) → ∈-++⁻ xs {ys} (∈-++⁺ˡ i) ≡ inj₁ i
+  ∈-++⁻∘++⁺ˡ {xs = xs} ys i = AnyP.++⁻∘++⁺ xs (inj₁ i)
+
+  ∈-++⁻∘++⁺ʳ : ∀ {x : A} xs {ys} (i : x ∈ ys) → ∈-++⁻ xs (∈-++⁺ʳ xs i) ≡ inj₂ i
+  ∈-++⁻∘++⁺ʳ xs i = AnyP.++⁻∘++⁺ xs (inj₂ i)
+
+  ∈-++⁺∘++⁻ : ∀ {x : A} xs {ys} (i : x ∈ xs ++ ys) → ∈-++⁺ (∈-++⁻ xs i) ≡ i
+  ∈-++⁺∘++⁻ xs i = AnyP.++⁺∘++⁻ xs i
+
+-- The *represented* Cartesian multicategories are those generated from a
+-- Cartesian category by `Rep`. A multimorphism of type `Γ ⇒ A` is a morphism
+-- of type `prod Γ ⇒ A` from the original category.
+
+module _ {C : Category o ℓ e} (cart : Cartesian C) where
+
+  open Category C
+  open Equiv
+  open HomReasoning
+  open Cartesian cart
+  open ×P C
+  open Prods cart
+
+  infix 4 _⇒Π_ _≈Π_
+
+  -- A map of type `X ⇒ prod Δ` is equivalently
+  -- a map of type `X ⇒ Y` for each Y ∈ Δ.
+  -- `X ⇒Π Δ` is exactly the latter.
+
+  _⇒Π_ : Obj → List Obj → Set (o ⊔ ℓ)
+  X ⇒Π Δ = Env (X ⇒_) Δ
+
+  _≈Π_ : ∀ {X Δ} (fs gs : X ⇒Π Δ) → Set (o ⊔ e)
+  _≈Π_ = [ _≈_ ]_≈ᵉ_
+
+  ⟨_⟩Π : ∀ {X Δ} → X ⇒Π Δ → X ⇒ prod Δ
+  ⟨_⟩Π {Δ = []} fs = !
+  ⟨_⟩Π {Δ = Y ∷ Δ} fs = ⟨ fs (here ≡.refl) , ⟨ fs F.∘ there ⟩Π ⟩
+
+  ⟨_⟩Π-resp-≈ : ∀ {X Δ} {fs gs : X ⇒Π Δ} → fs ≈Π gs → ⟨ fs ⟩Π ≈ ⟨ gs ⟩Π
+  ⟨_⟩Π-resp-≈ {Δ = []} ps = refl
+  ⟨_⟩Π-resp-≈ {Δ = Y ∷ Δ} ps =
+    ⟨⟩-cong₂ (ps .get (here ≡.refl)) ⟨ [ ps .get F.∘ there ] ⟩Π-resp-≈
+
+  projectΠ : ∀ {X Y Δ} (i : Y ∈ Δ) (fs : X ⇒Π Δ) → π[ i ] ∘ ⟨ fs ⟩Π ≈ fs i
+  projectΠ (here ≡.refl) fs = project₁
+  projectΠ (there i) fs =
+    trans assoc (trans (∘-resp-≈ refl project₂) (projectΠ i _))
+
+  ⟨π⟩Π : ∀ {Γ} → ⟨ π[_] {_} {Γ} ⟩Π ≈ id
+  ⟨π⟩Π {Γ} = uniqueness* {ys = Γ} λ i → trans (projectΠ i _) (sym identityʳ)
+
+  Rep : CartesianMultiCategory o ℓ e
+  Rep = record
+    { Obj = Obj
+    ; _⇒_ = λ Γ A → prod Γ ⇒ A
+    ; _≈_ = _≈_
+    ; id = π[_]
+    ; _∘_ = λ f σ → f ∘ ⟨ σ ⟩Π
+    ; equiv = equiv
+    ; ∘-resp-≈ = λ ff σσ → ∘-resp-≈ ff ⟨ σσ ⟩Π-resp-≈
+    ; identityˡ = λ {_} {_} {_} {i} → projectΠ i _
+    ; identityʳ = λ {Γ} → trans (∘-resp-≈ʳ (⟨π⟩Π {Γ})) identityʳ
+    ; identity² = λ {_} {_} {i} → projectΠ i _
+    ; assoc = λ {Γ Δ Θ A f σ τ} → begin
```

Comment:
So it turns out that inline equational proofs take a lot longer to typecheck (and produce bloated results); best to pull these out, either in a prepended `private` block or using a `where` block below.